### ISSUE TITLE
Accessibility: Progress indicator/loading UI for "Instant loading" configuration

### ISSUE THREAD

**Issue by netanel-haber-deci** [squidfunk/mkdocs-material#6087]
### Description

Right now, there is no indicator to the user that another page is loading when [Instant Loading](https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#instant-loading) is on. Sometimes pages are large and the experience is suboptimal.
It would be useful to be able to either get a progress indicator OOTB, or to be able to toggle/configure such a spinner. Maybe an overlay with a slight change in opacity for further enhancement would be useful as well.
If this already exists via a more low-level construct (specifying UI during a hook, etc.), the request is to add documentation for this use-case.
I quite like the [determinate linear indicator](https://m3.material.io/components/progress-indicators/specs#9194f993-31bb-44cb-8a46-f47be0245b91), shown at the very top of the page - this is pretty standard in plenty of websites.

### Related links

- [Requesting a change](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/)
- [Instant Loading](https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#instant-loading)
- [Material Progress Indicators](https://m3.material.io/components/progress-indicators/overview)

### Use Cases

This feature would increase accessibility can directly benefit all users of the instant loading feature, especially for navigations with large payloads.

### Visuals

- [Determinate linear indicator](https://firebasestorage.googleapis.com/v0/b/design-spec/o/projects%2Fm3%2Fimages%2Fl4yjc2jk-linear-progress-indicator-Determinate_3P.mp4?alt=media&token=f05ea51f-14e1-437b-988a-7a67c78b225e)

### Before submitting

- [X] I have read and followed the [change request guidelines](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/).
- [X] I have verified that [my idea is a change request and not a bug report](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#its-not-a-bug-its-a-feature).
- [X] I have ensured that, to the best of my knowledge, [my idea will benefit the entire community](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#benefit-for-the-community).
- [X] I have included relevant links to [the documentation](https://squidfunk.github.io/mkdocs-material/), related [issues](https://github.com/squidfunk/mkdocs-material/issues), and [discussions](https://github.com/squidfunk/mkdocs-material/discussions) to underline the need for my idea.

**Comment by squidfunk** [comment]
Thanks for suggesting. Yes, I think it makes sense to add something like this. In order to show progress, we'd have to patch our loading logic to use `XMLHttpRequest`, because it's the only way to track progress, but I think it makes sense to do.

Please clarify:

> Sometimes pages are large and the experience is suboptimal.

**Comment by netanel-haber-deci** [comment]
> Please clarify:
> 
> > Sometimes pages are large and the experience is suboptimal.

I meant that for slow internet connections, and large page loads, the navigation sometimes is slow to respond - and that is a suboptimal UX.

Regarding a forced move to `XMLHttpRequest` -  I haven't looked at this closely (I can try to get to it later) - but a cursory search found this: https://javascript.info/fetch-progress, which suggests it is possible to track download progress with native `fetch`, only upload progress would force the patch. Is there something I'm missing?

**Comment by squidfunk** [comment]
Thanks for researching – ah yeah, I was probably stuck in my head at tracking upload progress. However, I have trouble finding data on browser support for tracking download progress with fetch. Do you happen to have a source for that?

**Comment by netanel-haber-deci** [comment]
Okay - so I read a bit more. 
1. First of all, the basic [`body.getReader`](https://caniuse.com/mdn-api_readablestream_getreader) interface is widely supported, basically all active browsers aside from IE support it. [[Caveat for zero-copy BYOB reader?]](https://caniuse.com/mdn-api_webtransport_byob_readers)
2. Using it for progress has a big issue: Compressed responses. The article I linked to and other implementations rely on `Content-Length`. If the response is gzipped for example, the content length is the compressed length, while the read bytes are already decompressed. Unless servers also send an e.g. `x-uncompressed-length` header, it'll be non trivial to extract useful progress information.

Therefore, I think using  the `XMLHttpRequest` `onprogress` event is probably the way to go 🤷

**Comment by netanel-haber-deci** [comment]
Maybe someone with more experience with `XMLHttpRequest` vs `fetch` can weigh in here. 
If you do end up patching - do you think it would be a better idea to convert to XMLHttpRequest just for this specific case (i.e, only when `instant.navigation` is on + only for navigation xml downloads), or just use it for all requests, all the time?

**Comment by squidfunk** [comment]
If we decide to switch to `XMLHTTPRequest`, we should go all the way. We would need to adjust this function, and check all calls, because we need to adjust the options that we pass to the `request*` functions. `RequestInit` is part of  `fetch`:

https://github.com/squidfunk/mkdocs-material/blob/b77769009ece2326476d1a02d8fa04e501f73af0/src/templates/assets/javascripts/browser/request/index.ts#L50-L61

If you would like to work on this, we'd be happy to discuss implementation in a PR, but this feature definitely makes sense, so one way or the other, we should add it.

**Comment by squidfunk** [comment]
Added support for progress indicator in f8fd537b97f899d232ebfe2a6237869acea79124. Demo on `128kbit/s`:

https://github.com/squidfunk/mkdocs-material/assets/932156/9d930c18-b044-404c-b3e6-bd4ac6bd8211

Some design considerations:

- A new feature flag `navigation.instant.progress` and partial called `partials/progress.html` was added
- We decided not to use the `progress` element, as it looks different in all browsers and we would need to use pseudo elements to address it. Instead we just use a regular `div` with a `progress` landmark role. That should be sufficient for good accessibility. If not, let's improve it.
- We completely switched from `fetch` to `XHR`, as it makes no sense to maintain two ways of fetching data. My testing shows that all features still work as expected, including downloading of GitHub repository metadata. Tracking download progress is a big improvement and `fetch` has serious shortcomings here (as OP mentioned before).
- Progress immediately starts at 5% the moment the request is dispatched to denote that something is happening before content can be downloaded and progress can actually be tracked.
- The progress bar will show after a delay of 400ms, because on fast connections it's normally not necessary to show a progress bar. After 400ms it will fade in and show the progress until the request is complete. It will automatically fade out once 100% are reached. All of this (except for setting the value, which is done from javaScript, obviously) is implemented in CSS, and can thus be customized easily. If you want to lower the delay, just set `--md-progress-delay`. Our implementation can be inspected here:

    https://github.com/squidfunk/mkdocs-material/blob/9ac03cf806cd04a7530415ea4ed0af9288900004/src/templates/assets/stylesheets/main/components/_progress.scss#L27-L53

__We decided to add this feature to the community edition and not to Insiders, because we regard it as a big improvement for instant loading. It's now much more apparent what's happening in the background, especially on slow connections.__
